Make `IByteQueue.push` accept an `in` buffer argument (remake)

david-eckardt-sociomantic opened the pull request:
This method does not modify the input so it should accept const data.

Remake of #381.

david-eckardt-sociomantic committed:
Change queue methods to use `void[]` buffers with `const` where applicable

- Replace `ubyte[]` with `void[]`.
- For `push` and other methods that don't write to the buffer make it
  `const`.
- Make the buffer returned by `pop` and `peek` `const`.

mihails-strasuns-sociomantic reviewed (changes requested):
Migration notes and unit test please ;)

mathias-lang-sociomantic commented:
And rebase

david-eckardt-sociomantic commented:
Updated.

mathias-lang-sociomantic commented on relnotes/queue-push-const.migration.md:
```diff
@@ -7,3 +7,9 @@
+
+* `ocean.util.container.queue.model.IIUntypedQueue`
+
+The type of the `t` argument of `push` has been changed from `void[]` to
```
Indent

david-eckardt-sociomantic replied:
Oops

mathias-lang-sociomantic reviewed (approved):
LGTM

david-eckardt-sociomantic commented:
Updated.

nemanja-boric-sociomantic commented:
@mihails-strasuns-sociomantic ?

mihails-strasuns-sociomantic commented on src/ocean/util/container/queue/FixedRingQueue.d:
```diff
@@ -185,15 +187,16 @@ class FixedByteRingQueue : FixedRingQueueBase!(IByteQueue)
     ***************************************************************************/
 
-    bool push ( ubyte[] element )
+    bool push ( in ubyte[] element )
```
I guess only question remaining from my side is why this takes `in ubyte[]` and not `in void[]` - you wouldn't need to do all those casts in tests otherwise.

Considering this is a breaking change targeting v4.x.x, maybe it is time to go void? :)

david-eckardt-sociomantic replied:
I would like that, but it would be a rather intrusive change at a greater scope. It would in addition include
- `IByteQueue`
  - `ubyte[] push ( size_t size )`
  - `ubyte[] pop ( )`
  - `ubyte[] peek ( )`
- `NotifyingQueue`
  - `bool push ( size_t size, void delegate ( ubyte[] ) filler )`

It will most certainly break much more user code than just adding `in`. Right I'm torn here, `void[]` is much better IMHO.

david-eckardt-sociomantic replied:
We could deprecate `IByteQueue` and all other `*Byte*Queue`s in favour of `*Data*Queue` which is `const`-correct and uses `void[]`. That way the `const`-correctnes would even got in v3.x.x.

mihails-strasuns-sociomantic replied:
Sounds cool, though if we go this way, I would even leave `Byte*` part undeprecated first - it propagates up to swarm and forth which will annoy everyone again with dozens of unfixable deprecations.

david-eckardt-sociomantic replied:
There is a complication with class naming. The classes in question are `FlexibleFileQueue`, `NotifyingQueue` and `QueueChain`. These classes depend on `IByteQueue` and need `IDataQueue` clones. What would be the best procedure here? I would like to avoid to add a useless `Data` infix to the class names.

david-eckardt-sociomantic replied:
I guess I'm going to make it a breaking change in the next major release then.

mihails-strasuns-sociomantic replied:
I am utterly irresponsible about naming things, ask @gavin-norman-sociomantic :trollface: 

As for breaking change - I would aim for adding new stuff in v3.x.x, deprecating old one at some point after 4.0.0 and removing completely in v5.x.x. This way we will have simple transitional point of v4.0.0-v4.N.0 during which both old and new API is known to be available.

david-eckardt-sociomantic committed:
`SimpleStreamSerializer.writeData`: Accept `const` input buffer

Split the `transmitData` function body and move it into `readData` and
`writeData`, resp.

david-eckardt-sociomantic committed:
Add missing cast in unittest

david-eckardt-sociomantic committed:
Add alias to fix subclass overload set

This is needed for `T == char[]` in the unit test, which creates the
overload set of `NotifyingQueue!(char[]).push(ref char[])` with its super
class `NotifyingByteQueue.push(in void[])` method.

david-eckardt-sociomantic committed:
Small type fixes

david-eckardt-sociomantic committed:
`NotifyingQueue.push`: Don't use `copy`

`target` should have the required length already and should not be resized.

david-eckardt-sociomantic commented:
Updated, in addition
- changing `ubyte[]` to `void[]`,
- making `push` and `peek` return `const` arrays,
- `const`-correcting the `FlexibleByteQueue.save` delegate.

mihails-strasuns-sociomantic commented:
With next ocean v4.0.0 alpha will be interesting to check how this affects swarm.
